Initial changes to make the Today Home Widget customizable.

diegoreymendez opened the pull request:
First attempt at making our Today Home Widget customizable.

In this PR you can:

- Customize the widget to select which site to show the stats for.
- See the site stats in the widget.

This PR also adds a new target (the Intents Extension) together with provisioning profiles and App ID.

## About the number of lines changed:

The PR is quite large, with around 2000 lines added.  Most of these, though, are changes to add the new intents target and the certificates.  The code changes aren't that many.

## Known limitations:

- This PR doesn't change the stats refresh logic so the info may not be up to date at all times.
- Remote icons for sites aren't loading yet: Apple's default mechanism for downloading remote images in the intents extension is not working at all, and I'm a bit stuck trying to figure out the best solution.
- It seems we're showing hidden sites too.  We should probably filter the list of sites shown to visible ones.

## To test:

1. Long tap on the widget.
2. Tap "Edit Widget"
3. Tap "Choose" in the "Site" row.
4. Select a site from your list.

Make sure the site you selected is properly loaded in the widget.

## PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

diegoreymendez committed:
WIP

diegoreymendez committed:
Merges the latest from the feature branch.

diegoreymendez committed:
rake lint:autocorrect

diegoreymendez committed:
Fixes a build issue.

diegoreymendez committed:
Added provisioning profiles and code to show sites in the Home widget configuration.

diegoreymendez committed:
Added several configuration schemes.

jkmassel committed:
Set up Intents Signing

diegoreymendez committed:
Adding images to the site selection in the home widget.

diegoreymendez committed:
Removes some unnecessary comments.

diegoreymendez committed:
Adds some improvements.

diegoreymendez committed:
Some formatting cleanup.

frosty commented on WordPress/Info.plist:
```diff
@@ -8,96 +8,85 @@
 	<key>CFBundleExecutable</key>
 	<string>${EXECUTABLE_NAME}</string>
-	<key>CFBundleIcons~ipad</key>
+	<key>CFBundleIcons</key>
```
I _think_ it's just moving items in the plist around automatically, but I'm not sure these icon changes really belong in here?

diegoreymendez replied:
I've rolled back all changes to this Info.plist since none of these changes done manually.  I believe the one to trigger all the changes was the `NSUserActivityTypes` one.

I've re-tested my PR without that change and it seems to work just fine.  Can I ask you to double check this @frosty ?

frosty commented on WordPress/WordPress.xcodeproj/xcshareddata/xcschemes/WordPressNotificationServiceExtension.xcscheme:
```diff
@@ -0,0 +1,96 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Any idea why these schemes for other targets were added?

diegoreymendez replied:
I re-ordered the schemes a bit, and I'm not sure why but that seems to have shared the schemes as well.  I un-shared the schemes and the files are now gone.

diegoreymendez committed:
Shows the selected site in the widget.

diegoreymendez committed:
rake lint:autocorrect

diegoreymendez commented:
Update: I've pushed more changes into this PR so that the site you select is shown in the widget.

frosty commented on WordPress/Info.plist:
```diff
@@ -253,10 +253,22 @@
+	<string>To add photos or videos to your posts.</string>
+	<key>NSUserActivityTypes</key>
+	<array>
+		<string>SelectSiteIntent</string>
```
Should this be reverse-dns style like the others?

diegoreymendez replied:
Changes reverted... read my comment above.

frosty commented on WordPress/WordPressHomeWidgetToday/WordPressHomeWidgetToday.swift:
```diff
@@ -67,12 +68,102 @@ private extension Provider {
+
+    private func widgetData(for siteID: String) -> HomeWidgetTodayData? {
+        // TODO - TODAYWIDGET: we might change this, but for now an ID equal to zero should not return any valid data
+        //HomeWidgetTodayData.read()?[siteID ?? 0]
```
Can we remove this line?

diegoreymendez replied:
Done.

frosty commented on WordPress/WordPressHomeWidgetToday/WordPressHomeWidgetToday.swift:
```diff
@@ -67,12 +68,102 @@ private extension Provider {
+              let siteIdentifier = site.identifier,
+              let widgetData = widgetData(for: siteIdentifier) else {
+
+            completion(Self.staticContent)
```
Yes, presumably this should show 0 values and some kind of error if there's no site / data? (this can be for a future PR!)

diegoreymendez replied:
I'll leave this for future evaluation, since I'm not sure it's safe NOT to call the completion closure.

Personally I'd expect the `completion` closure to always be called, one ways or another, since it doesn't specifically signal a "success" situation, but the end of an operation.

Gio2018 replied:
as far as I understood, `getSnapshot` is used in transient situations, like when the user is installing the widget (so it's sort of a preview): in this case I think it's good to show some local data if we have them already. If we don't, we should probably try to understand what kind of information we want to show in that place (the current `staticContent` needs to be changed anyway, or we should use a different view entirely)

diegoreymendez replied:
Marking this as resolved for now as there's no further action required.

frosty commented on WordPress/WordPressIntents/IntentHandler.swift:
```diff
@@ -0,0 +1,73 @@
+        return sitesDataProvider.defaultSite
+    }
+
+    func resolveSite(for intent: SelectSiteIntent, with completion: @escaping (SiteResolutionResult) -> Void) {
```
I think this is primarily intended when using intents with SiriKit. For example, if a user spoke the name of a site to use, you would have to try and match it to one of their sites in the app – you could then pass back either a successful match or a failure. See https://developer.apple.com/documentation/sirikit/resolving_and_handling_intents/resolving_the_parameters_of_an_intent

diegoreymendez replied:
Updates this comment to reflect I probably will need to implement that method (in a future PR).

frosty reviewed (commented):
Functionally, this is working as expected – I was able to see my sites in the list and select one, and that selection was saved. I've left some code level comments, and I have a few other things:

- I'm not sure we need the site icon on the configuration view – particularly if a site has a longer title, we might not have space anyway 🤔
<img width="362" alt="image" src="https://user-images.githubusercontent.com/4780/100446410-a96b0680-30a6-11eb-9e27-36ed545af856.png">

- The sites on the list don't appear to be in any specific order, and it seems to be randomized each time
- In the main app, we show the site's domain as a subtitle on the cell – is that possible here?
- Do we need the IntentsUI target for this? I was under the impression that was only for use with Siri or Maps, but I may not understand it correctly.

diegoreymendez committed:
Merges the latest from the feature branch.

diegoreymendez commented:
I'm temporarily converting this back to a draft and removing the reviewers, as I fix the merge issues.  Please hold any reviews.

diegoreymendez committed:
Fixes all current merge conflicts.

diegoreymendez committed:
Rolls back some unnecessary changes.

diegoreymendez commented:
The merge conflicts are resolved, but I'll keep this as draft while I address @frosty 's feedback and also check why it is that site icons are NOT loading in the site selection screen for the widget.

diegoreymendez commented:
Gio - this is ready for review.  I'll be addressing @frosty 's feedback in the meanwhile.

diegoreymendez committed:
Merges the latest from the feature branch.

diegoreymendez committed:
Rolls back the undesired changes to Info.plist.

diegoreymendez committed:
Removed some no-longer-necessary comments.

diegoreymendez committed:
Updates a comment. Changed the name of a file that was unclear.

diegoreymendez committed:
Removed the WordPressIntentsUI target, since it wasn't being used.

diegoreymendez committed:
Unshared several schemes that were shared for no-known reasons.

diegoreymendez committed:
Updates the fastlane FastFile to remove a target that doesn't exist anymore.

diegoreymendez commented on WordPress/WordPress.xcodeproj/xcshareddata/xcschemes/WordPressIntents.xcscheme:
```diff
@@ -0,0 +1,96 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Folks, I'm not sure if this scheme should be shared or not.  IMHO there's no reason why schemes shouldn't be shared but if either of you knows better, can you let me know?

PS: Regardless of my doubt, I still haven't shared the others on purpose.  Not sure what happened there!

frosty replied:
I'm not sure either, although it appears we don't currently share any schemes (although it's also not in our gitignore)... @jkmassel any ideas?

diegoreymendez committed:
Moves the default site logic to the intents extension.

diegoreymendez committed:
Handles the default site through the intents extension.

diegoreymendez committed:
Manually merges some changes to WordPressHomeWidgetToday that had been lost.

diegoreymendez committed:
Removes an unnecessary todo.

diegoreymendez commented:
Just for context, the rendering issue in the main Widget customization screen seems to be [a bug](https://twitter.com/simonbs/status/1309235571093536770?s=20).

~I'll try to work around it.~  That just didn't work.  I'm turning off site icons for now.

diegoreymendez committed:
Removed images from the widget site picker. Added domain as subtitle.

diegoreymendez commented:
This is ready for a full review.  Please make sure to test the Widget is updating properly, following [the steps described here](https://github.com/wordpress-mobile/WordPress-iOS/pull/15391).

diegoreymendez commented:
Replying to @frosty :

> I'm not sure we need the site icon on the configuration view – particularly if a site has a longer title, we might not have space anyway 🤔

I just removed the images for now since I couldn't fix the bug you reported.  It seems to be an SDK bug.

> The sites on the list don't appear to be in any specific order, and it seems to be randomized each time

Will work on this in another PR.  We also have to remove the hidden sites.

> In the main app, we show the site's domain as a subtitle on the cell – is that possible here?

Done.

> Do we need the IntentsUI target for this? I was under the impression that was only for use with Siri or Maps, but I may not understand it correctly.

Removed.

Gio2018 commented on WordPress/Classes/Stores/StatsInsightsStore.swift:
```diff
@@ -1014,6 +1014,7 @@ private extension InsightStoreState {
+                                                           iconURL: blog.icon ?? oldData.iconURL,
                                                            url: blog.url ?? oldData.url,
                                                            timeZone: blogService.timeZone(for: blog),
                                                            date: Date(),
```
nit: the indentation here needs fixing, but I think that comes from the code I merged in earlier 😀

diegoreymendez replied:
Fixed.

Gio2018 commented on WordPress/WordPressHomeWidgetToday/WordPressHomeWidgetToday.swift:
```diff
@@ -1,106 +1,129 @@
+        .supportedFamilies([.systemSmall, .systemMedium])
+    }
+
+    /*
```
can we remove the commented out code? Seems that we won't need it anyway

diegoreymendez replied:
Definitely.  I didn't notice this was still there.

Gio2018 reviewed (approved):
Looks good to me! I installed the widget, chose different sites and observed stats updating for that site either when the app received new stats or when the widget refreshed the timeline (had to tweak the refresh interval for that).

just left a few minor comments in the code, otherwise :shipit:!

diegoreymendez committed:
Fixes the indentation of some lines of code.

diegoreymendez committed:
Removes some commented out code.

diegoreymendez commented:
Hey @frosty , @Gio2018 - Since this is targetting a feature branch and Giorgio already approved, I'm moving ahead to merge.

Please feel free to add comments on this PR and I'll address them in a separate PR.

@jkmassel - Please let me know what you think about this: https://github.com/wordpress-mobile/WordPress-iOS/pull/15397#discussion_r532963217
